oauth2/oauth2-resource-server/src/main/java/org/springframework/security/access/expression/AuthenticationResolver.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.access.expression;
+
+import org.springframework.security.core.Authentication;
+
+import java.util.function.Function;
+
+/**
+ * A contract for indicating work to be done that relies on the present
+ * {@link Authentication} object
+ *
+ * @author Josh Cummings
+ * @since 5.1
+ */
+public interface AuthenticationResolver {
+	<OUT> OUT doWithAuthentication(Function<Authentication, OUT> todo);
```

Comment:
I'm not sure I understand the reason for `Function` as the input type. Can you please clarify?
I was thinking the contract would be something like 

`Authentication resolve()`

OR

`<T extends Authentication> T resolve(Class<T> type)`

FYI, for reactive it would need to return a `Mono`, e.g. `Mono<Authentication> resolve()`

Reply:
I'm wondering if a contract like this would hide the "reactiveness" from the interface. The `Function` is what the caller would like the resolver to do. The resolver supplies the `authentication` in whatever way is cares to. 

Thus, someone in a reactive context would call

```
Mono<WhateverGetsDerivedFromTheOperation> result = resolver.doWithAuthentication((authentication) -> { return theReactiveThingINeedToDoToGetMyResult(); });
```

I need to get more familiar with our reactive library to know whether that would feel funny to users, but that is how I used it in `OAuth2AccessTokenClaimsExpressionMethods`.

Reply:
Thinking about this more, I'm guessing that a user is going to want to receive a `Mono<Authentication>` in their function, too, just like they'd want to receive a `Mono<Authentication>` as a return type. Curses.

Reply:
I'd prefer our logic be in a method signature like:

```java
public class OAuth2Expressions {
    public boolean hasScope(Authentication authentication, String scope) {

    }
    ...
}
```

This has some benefits:
* Our logic is more consistent with how things work now. For example, `PermissionEvaluator` API accepts the `Authentication` as an argument.
* It is more flexible and will work in a reactive environment. As it stands nothing within the `OAuth2AccessTokenClaimsExpressionMethods` API can resolve an `Authentication` using the Reactor Context because the return type of the methods do not return `Mono` or `Flux`. This means we would either need to create a new instance of `OAuth2AccessTokenClaimsExpressionMethods` for every request which is not good for garbage collection.

We could start by requiring the user explicitly providing the `Authentication` object as a return value. To make it easier we have a few options:

The first option is to create a separate object that resolves the `Authentication` for the user from the `SecurityContext`. For example:

 public class OAuth2ExpressionsBean {
    private final OAuth2Expressions oauth2 = new OAuth2Expressions();
    public boolean hasScope(String scope) {
        return oauth2.hasScope(SecurityContextHolder.getContext().getAuthentication(), scope);
    }
    ...
}

The advantage to this approach is all our logic is reusable within a reactive world. The downside to this approach is that we must duplicate every method.

Alternatively, and ideally we could plug in a custom `MethodResolver` and a custom `MethodExecutor` to automatically resolve the `Authentication` perhaps based on an annotation like `@SecurityContextAuthentication`. This would also benefit users who have their own custom Beans who want the `Authentication` to automatically be resolved. An example of our bean would be:

```java
public class OAuth2Expressions {
    public boolean hasScope(@SecurityContextAuthentication Authentication authentication, String scope) {

    }
    ...
}
```

Reply:
@rwinch, thanks for the feedback. 

While it is a bit duplicative, I like how you've separated the concern with your first solution.

I initially researched a way to do as you said with a `MethodResolver`, but couldn't get my head around how SpEL would take an expression like `hasScope('read')` and know to invoke `hasScope(Authentication, String)`. Based on your comment, though, I suppose SpEL may have some way to support that?

Reply:
> I initially researched a way to do as you said with a MethodResolver, but couldn't get my head around how SpEL would take an expression like hasScope('read') and know to invoke hasScope(Authentication, String). Based on your comment, though, I suppose SpEL may have some way to support that?

Yes. This is quite possible using a custom `MethodResolver` and custom `MethodExecutor`

Reply:
I also like:

```
public class OAuth2Expressions {
    public boolean hasScope(Authentication authentication, String scope) {

    }
    ...
}
```

But I don't feel `Authentication` is required in each method. We could define an interface:

```
public interface OAuth2Expressions {
    public boolean hasScope(String scope);
    
    public boolean hasAnyScope(String... scope);
    ...
}
```

And the implementations can be constructed with an `Authentication` via it's constructor. The `Authentication` would first be resolved in the Servlet or Reactive environment and than the implementation of `OAuth2Expressions` would be constructed with the resolved `Authentication`.